plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php

```diff
@@ -799,6 +796,33 @@ public function build_frontend_query( $query, $block, $page ) {
 		);
 	}
 
+	/**
+	 * Return a custom query based on attributes, filters and global WP_Query.
+	 *
+	 * @param array $parsed_block  A representative array of the block being rendered. See WP_Block_Parser_Block.
+	 *
+	 * @return array
+	 */
+	public function set_fallback_query_id( $parsed_block ) {
+
+		$block_name = $parsed_block['blockName'] ?? null;
+
+		if ( 'woocommerce/product-collection' !== $block_name ) {
+			return $parsed_block;
+		}
+
+		if ( empty( $parsed_block['attrs']['queryId'] ) ) {
+
+			// Build a consistent query ID based on the query parameters.
+			$hashed_args = $parsed_block['attrs']['query'] ?? array();
+			unset( $hashed_args['paged'] );
+
+			$parsed_block['attrs']['queryId'] = md5( wp_json_encode( $hashed_args ) );
+		}
```

Comment:
Wow! This is a clever solution. I like it. There is one small problem. `queryId` should be a valid number, but `md5()` returns a string. Maybe we can use `crc32()` instead? i.e. 🤔
```
$parsed_block['attrs']['queryId'] = abs( crc32( wp_json_encode( $hashed_args ) ) );
```

Another problem will be when `$parsed_block['attrs']['query']` is same for two Product Collections on same page. This is an edge case but I believe this can still happen. For this to happen:
- Two Product Collections on same page
- Both have same query params
- Both doesn't have any `queryId`

To make it a bit better, probably we can use `$parsed_block['attrs']` instead of `$parsed_block['attrs']['query']` to generate the `queryId` hash. This won't fix the edge case but will make it a bit better.

I am not 100% sure if we should address this edge case. @kmanijak What do you think about this edge case? Should we address it?

Reply:
FYI, I see the forced page reload in my branch versus the trunk as well

Reply:
Update, I _think_ it's that the [`enable_client_side_navigation` is looking for the `queryId` in the stashed `$this->parsed_block`](https://github.com/woocommerce/woocommerce/blob/1253b4ba54463c43ca5878ec5b02b694d646b36c/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L346)

Reply:
Update, I am seeing the `wp-block-query-pagination` links are missing their interactivity directives now (versus where it's working with a `queryId` from the editor). Unsure why.

Reply:
> Does queryId have to be an integer? Could we run abs() or intval() on the md5()?

Yes, `queryId` needs to be an integer for the client-side pagination to work correctly. Using `abs()` or `intval()` on `md5()` would work technically, but it would truncate most of the hash value since `md5()` produces a 32-character string.

That's why I suggested using `crc32()` which produces an integer hash directly. However, as noted in the PHP docs, we should wrap it in `abs()` to handle negative values on 32-bit systems: